### ISSUE TITLE
SpaceFN is a bit too slow

### ISSUE THREAD

**Issue by toneman77** [rvaiya/keyd#34]
First of all: as a user with a QMK keyboard that also has to regularly type on a laptop your keyd is a blessing.

As a diehard vim guy I quickly was able to translate the most important stuff of my qmk setup to keyd:
```
# Turns capslock into an escape key when pressed and a control key when held.
capslock = overload(C, esc)

# Remaps the escape key to capslock
# esc = capslock

# space when tapped, spacefn when held
space = overload(spacefn, space)
tab = overload(tabfn, tab)

[spacefn]
h = left
j = down
k = up
l = right

[tabfn]
h = home
j = pagedown
k = pageup
l = end
```
- caps as ctrl/esc works like a charm, bye xcape
- initial tests with SpaceFn and TabFn work: nice, bye xmodmap+setxkbmap

Then I started to write an email and realizedthatthespace keyddoesnt work asreliable :/
My typing speed is non-spectacular (75-80 wpm according to monkeytype.com) but a space doesnt come through two thirds of the time.

Is there anything I can do to make this more accurate? Did I configure something wrong?

**Comment by rvaiya** [comment]
My guess would be it has to do with timeouts. keyd uses sequence rather than time based tap detection which works well as long as you release the first key before pressing the second. The end result is entiresly deterministic but can cause confusion in the case of certain keys. Disambiguating `<key1 down> <key2 down> <key1 up> <key2 up>` depends on the user's typing habits and requires different configurable timeouts to get right. I am considering adding support for this but it might break certain assumption currently made in the core logic. What timeouts do you use on QMK? See also #15.

**Comment by toneman77** [comment]
I didnt adjust them for my layouts see https://github.com/qmk/qmk_firmware/tree/master/keyboards/gh60/satan/keymaps/iso_split_rshift (I switched keyboards since then, but the keymap layout is more or less similar)
I'll have to research a bit on what those defaults are

**Comment by dit7ya** [comment]
I just installed Keyd now (have been a happy user of KMonad so far) and I am facing the same problem.

**Comment by slakkenhuis** [comment]
Checking in with the same issue. I adore the design of keyd, it's just so *clean* --- this issue is the one thing holding me back from switching. Thanks so much for your work!

Somewhat surprising: I now use xcape to set esc/enter on tapping left alt/right alt respectively, but even without setting a `-t` timeout on `xcape Alt_L=Escape;Alt_R=Return`, the keypresses come through just fine. Not so with:

```
leftalt = overload(A, escape)
rightalt = overload(A, enter)
```

My layman's intuition is that it has to do with xcape being meant only for *modifier* keys. Suppose you type `<space down> <ralt down> <space up> <ralt up>`, it's pretty clear that you meant to do `space enter`, not `A-space`. The ambiguity only comes when `space` is also possibly a modifier when held. Perhaps a simpler solution, without timeout-based logic, would be to make it so that non-overloaded keys X that overlap with the later keypress of an overloaded key Y do not affect the registering of a tap of Y?

Not to say that timeout-logic would not also be useful, but I think the above would address the core issue for me.

**Comment by rvaiya** [comment]
> Not to say that timeout-logic would not also be useful, but I think the above would address the core issue for me.

I am in the process of making some internal changes in preparation for a 2.0 release. I intend to add support for some timeouts shortly after that has been released.

> Suppose you type <space down> <ralt down> <space up> <ralt up>, it's pretty clear that you meant to do space enter, not A-space.

This is indeed a bug. I had accounted for this ambiguity this in an earlier iteration of keyd but neglected to account it in the current one. It was trivial enough to warrant a bugfix on the v1 branch (currently master). Give the latest commit a try. 

If you still have issues then you may want to wait for timeout support, though I would recommend switching to 2.0 once it is released anyway, as it makes some minor changes to the config format.

**Comment by rvaiya** [comment]
Leaving this open until @dit7ya and @toneman77's confirm that their issue is fixed. Their use cases may still benefit from timeouts.

**Comment by slakkenhuis** [comment]
That was fast :) Glad it was easy. And I can confirm that the latest commit resolves my issue. Many thanks!

**Comment by slakkenhuis** [comment]
As for benefiting from timeouts --- indeed, I think that timeouts would still be necessary for overloading a key like space.

For my `leftalt`/`escape` and `rightalt`/`enter` overloading, the logic works fine without timeouts, since it doesn't occur often that I'll press another key before having released `escape` or `enter`.

But I've also been trying to overload `space`, and that's more problematic, as it occurs within words. To make that work, a timeout would be necessary that:

1. ... registers as a tap if held for less than *x* ms, *even* if it looks like another key is chorded alongside it. This would be like the 'alone' parameter suggested in #69 (I suppose the remainder of this issue would be a duplicate of that issue, anyway).

2. ... registers the tap on key press, rather than key release. That is, `<space down> <a down> <a up> <space up>`, with a less-than-*x* ms delay between the two keypresses, would show up as `space a`, not as `a space`. This means that the timed key would have to *queue* subsequent keypresses until either the timeout or the key release. After all, only at that point can it be disambiguated whether it was a tap or a hold.

I'm not sure whether such complex behaviour would lie within the admirably limited scope of `keyd`. It would be neat if it does, as I would *really* like to use `space` as a modifier. On the other hand, in other key remappers, I remember giving up on it because I found it bothersome that the space-tap would (necessarily) no longer appear instantaneously.

I'm curious about your thoughts/experiments with this.

For reference, [kmonad](https://github.com/kmonad/kmonad/blob/master/doc/quick-reference.md#tap-buttons) has `tap-hold` for this logic, while the current behaviour corresponds with their `tap-next`. I'm going to experiment with that for a bit.

**Comment by lemnos** [comment]
> As for benefiting from timeouts --- indeed, I think that timeouts would still be necessary for overloading a key like space.

Agreed.

> I found it bothersome that the space-tap would (necessarily) no longer appear instantaneously.

This is part of the reason a ended up axing timeouts from my earlier prototypes. I was originally keen on making them work, and while I made some progress, in the end I couldn't find a solution that seemed to *completely* eliminate ambiguity without either making keys feel unresponsive or requiring me to retrain myself to accommodate timeouts.

I came to the conclusion that time based ambiguity is a bad idea because the (my?) mind naturally associates action with consequence and time adds a third dimension. This is why `af` as an intention can yield the sequence `<a down> <f down> <f up> <a up>`, because your mind only associates actuation of `a` with the down stroke. 

Even if you immediately treat all other permutations of `<f down> <f up> <a up>` as `af` (which is a useful micro optimization) you are left with the inherently ambiguous sequence above, necessitating the timeout you have proposed. Your mind now must now insert a timeout into the sequence of instructions it produces to communicate your intent to the machine. That is, instead of immediately converting a goal like 'move one page down' into a discrete action like `control+f` you must pause mid stroke to disambiguate your intent.

what was before `<control down> <f down>` now becomes

`<a down> <timeout> <f down>`

which is distinct from

`<a down> <f down>`


This initially seems like a small price to pay, but adds a non trivial amount of friction to the user's interaction with the machine, slowly corroding their spirit ~~and potentially culminating  in a homicidal rage~~. 

Even with training, how reliably can you distinguish between `af` and `control-f`? Some people try and strategically overload keys within their letter layout to minimize the likelihood of certain permutations, but it is impossible to completely eliminate them, and in the end you just give yourself modifier PTSD.  Of course devotees of this ~~perverse~~ approach will insist that [it causes them no grief](https://cdn.discordapp.com/attachments/759649828136812596/921665889949069352/this-is-fine.0.jpg).

There is an interesting [blog post](https://pavelfatin.com/typing-with-pleasure/#human-side) about input latency by Pavel Fatin that you might find interesting if you haven't already seen it. His conclusions seem to accord with my intuition and the (unscientific) experiments I conducted. I have become convinced that trying to overload certain keys (as tempting as it may be) is ~~the province of swivel eyed loons~~ futile.

Having said all that, I don't want to be too prescriptive. Just because *I* found the feature unusuable doesn't mean that others might not profit from it.  Finally, there is always the possibility that I need to [be a man](https://www.youtube.com/watch?v=-qtrAMK7_Qk) and develop the requisite dexterity. And so, with that in mind I will probably add some support for timeouts to the 2.0 release.

TLDR;

Coming soon to a keyd near you.

**Comment by slakkenhuis** [comment]
> This initially seems like a small price to pay, but adds a non trivial amount of friction to the user's interaction with the machine

Yeah. My initial expectation was that there would be a 'natural' pause that occurs when you partner an overloaded modifier with another key — a pause that is not present when you simply tap it. While this did prove somewhat true (the sweet spot sat at around 150ms for me in my tests with `kmonad`), it does not eliminate the latency issue outlined in the article you linked (which I hadn't come across before).

> Having said all that, I don't want to be too prescriptive. Just because I found the feature unusable doesn't mean that others might not profit from it. Finally, there is always the possibility that I just need to be a man and develop the requisite dexterity. And so, with that in mind, I will probably add some support for timeouts to the 2.0 release.

Commendable flexibility. But I tend to agree with your assessment. It's probably not a good idea to use the presence of a timeout to communicate intent.

That said, I do still see use for timeouts for the opposite --- that is, not to *generate* a tap, but to *suppress* a hold. That needs no queue and could weed out some accidental keypresses that I'm seeing on my setup, albeit rarely. So...

> Coming soon to a keyd near you.

Yay!

Anyhow, thanks, this is illuminating.

**Comment by rvaiya** [comment]
Overload now accepts an optional third parameter which corresponds to a hold timeout in milliseconds. Give it a try and let me know how it goes. There are a number of other changes in the latest commit, so you may need to make some additional modifications to your config file (see the changelog).

**Comment by slakkenhuis** [comment]
Wow, at first sight it works --- surprisingly well. To the extent that I'm considering overloading space, after all. I thought that one was a lost cause (it might still be, of course). Great work! I'll let you know how it goes as I'm actually using it.

**Comment by rvaiya** [comment]
Thanks. I would also appreciate feedback from @toneman77  and @dit7ya.

**Comment by toneman77** [comment]
hey. 
sadly I wont be near my laptop until monday. then I can test it

**Comment by rvaiya** [comment]
No worries, respond at leisure.

**Comment by slakkenhuis** [comment]
Feedback upon extended usage: as expected, the functionality does fail to completely escape the pitfalls you described earlier. That said, as I mentioned, it's far more usable than kmonad's, if only because it hasn't bothered me to have a timed overload active when I'm not using it. It's a useful feature.

There is another problem with it, but it is sufficiently separate to have warranted opening another issue at #81 so as to avoid cluttering this one.

**Comment by lemnos** [comment]
Thanks for the feedback :). I was careful to keep the default behaviour of overload sequence based for this reason (and to avoid breaking old configs).

**Comment by rvaiya** [comment]
Closing due to inactivity. @toneman77 and @dit7ya feel free to reopen if your issue remains unsolved.

**Comment by vaalkor** [comment]
Just a note for anyone in the future who was a bit confused like me when they tried to add the timeout to `overload` and it didn't work: The overload that takes a timeout is called `[overloadt](https://github.com/rvaiya/keyd/blob/master/docs/keyd.scdoc#L696)`. 

I've just tried it and it works well for me considering I only rebind some keys they are not very time critical (arrow keys, enter, delete and a few others).

Thanks for writing great software people :)
